src/server.rs

```diff
@@ -58,65 +54,567 @@ use protocol::{
+
+    /// Zero and return stats about the cache.
+    fn zero_stats(&self) -> CacheStats {
+        *self.stats.borrow_mut() = ServerStats::default();
+        self.get_stats()
+    }
+
+
+    /// Handle a compile request from a client.
+    ///
+    /// This will either start compilation and set a `CompileStarted`
+    /// response in `res`, or set an `UnhandledCompile` response in `res`.
+    fn handle_compile(&self, mut compile: Compile)
+                      -> Box<Future<Item = SccacheResponse, Error = io::Error>>
+    {
+        let exe = compile.take_exe();
+        let cmd = compile.take_command().into_vec();
+        let cwd = compile.take_cwd();
+        // See if this compiler is already in the cache.
+        match self.compiler_info_cached(&exe) {
+            Some(info) => {
+                trace!("compiler_info cache hit");
+                future::ok(self.check_compiler(info, cmd, cwd)).boxed()
+            }
+            None => {
+                trace!("compiler_info cache miss");
+                // Run a Task to check the compiler type.
+                let exe2 = exe.clone();
+                let me = self.clone();
+                let creator = me.creator.clone();
+
+                let info = self.pool.spawn_fn(move || {
+                    Ok(get_compiler_info(creator, &exe2))
+                });
+
+                let me = self.clone();
+                Box::new(info.map(move |info| {
+                    me.cache_compiler_info(exe, &info);
+                    me.check_compiler(info, cmd, cwd)
+                }))
+            }
+        }
+    }
+
+    /// Look up compiler info from the cache for the compiler `path`.
+    fn compiler_info_cached(&self, path: &str) -> Option<Option<Compiler>> {
+        trace!("compiler_info_cached");
+        match metadata(path) {
+            Ok(attr) => {
+                let mtime = FileTime::from_last_modification_time(&attr);
+                match self.compilers.borrow().get(path) {
+                    // It's a hit only if the mtime matches.
+                    Some(&Some(ref c)) if c.mtime == mtime => Some(Some(c.clone())),
+                    // We cache non-results.
+                    Some(&None) => Some(None),
+                    _ => None,
+                }
+            }
+            Err(_) => None,
+        }
+    }
+
+    /// Store `info` in the compiler info cache for `path`.
+    fn cache_compiler_info(&self, path: String, info: &Option<Compiler>) {
+        self.compilers.borrow_mut().insert(path, info.clone());
+    }
+
+    fn check_compiler(&self,
+                      compiler: Option<Compiler>,
+                      cmd: Vec<String>,
+                      cwd: String)
+                      -> SccacheResponse {
+        let mut res = ServerResponse::new();
+        match compiler {
+            None => {
+                debug!("check_compiler: Unsupported compiler");
+                self.stats.borrow_mut().requests_unsupported_compiler += 1;
```

Comment:
Having to put `borrow_mut()` everywhere here is not wonderful. I can live with it for now but I'll have to think about the design some more at another point.